src/backend/columnar/columnar_tableam.c

```diff
@@ -983,8 +1070,149 @@ columnar_index_build_range_scan(Relation heapRelation,
 								void *callback_state,
 								TableScanDesc scan)
 {
-	ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-					errmsg("indexes not supported for columnar tables")));
+	/*
+	 * TODO: Should this function call pgstat_progress_update_param in
+	 * somewhere as heapam_index_build_range_scan ?
+	 */
+
+	if (start_blockno != 0 || numblocks != InvalidBlockNumber)
+	{
+		/* TODO: make sure that this is only valid for BRIN indexes */
+		ereport(ERROR, (errmsg("BRIN indexes on columnar tables are not supported")));
+	}
+
+	if (scan || indexInfo->ii_Concurrent)
+	{
+		/*
+		 * "scan" can be "already initialized" iff we are building the index
+		 * concurrently. We already don't allow concurrent index builds but
+		 * be on the safe side.
+		 */
+		ereport(ERROR, (errmsg("concurrent index builds are not supported "
+							   "for columnar tables")));
+	}
+
+	/*
+	 * Prepare for scan of the base relation. In a normal index build, we use
+	 * SnapshotAny because we must retrieve all tuples and do our own time
+	 * qual checks (because we have to index RECENTLY_DEAD tuples). In a
+	 * concurrent build, or during bootstrap, we take a regular MVCC snapshot
+	 * and index whatever's live according to that.
+	 */
+	TransactionId OldestXmin = InvalidTransactionId;
+
+	/*
+	 * We already don't allow concurrent index builds so ii_Concurrent
+	 * will always be false, but let's keep the code close to heapAM.
+	 */
+	if (!IsBootstrapProcessingMode() && !indexInfo->ii_Concurrent)
+	{
+		/* ignore lazy VACUUM's */
+		OldestXmin = GetOldestXmin(heapRelation, PROCARRAY_FLAGS_VACUUM);
+	}
+
+	/*
+	 * For serial index build, we must begin our own heap scan in this case.
+	 * We may also need to register a snapshot whose lifetime is under our
+	 * direct control.
+	 */
+	Snapshot snapshot;
+	bool need_unregister_snapshot = false;
+	if (!TransactionIdIsValid(OldestXmin))
+	{
+		snapshot = RegisterSnapshot(GetTransactionSnapshot());
+		need_unregister_snapshot = true;
+	}
```

Comment:
add a test for that if possible (that would automatically test the line with `UnregisterSnapshot(snapshot)` too)

Reply:
I couldn't come up with a test case for that. However, after implementing `CONCURRENTLY` support, we can easily test this part too.